### ISSUE TITLE
Keep scope of dependencies to this repository

### ISSUE THREAD

**Issue by Silvanoc** [containers/image#552]
`skopeo` is being used to test `containers/image`, but when building `skopeo` the `vendor.conf` of `containers/image` is not being taken into account. Therefore adding new dependencies to `containers/image` breaks the tests until the same dependencies have been also added to `skopeo`. [This pipeline](https://travis-ci.org/containers/image/jobs/470428966) of PR #546 is showing this issue.

I'm not that experienced at Go programming, but I'd expect dependencies of `containers/image` to be needed only in this repository an not "radiating" to neighbor repositories.

**Comment by vrothberg** [comment]
@Silvanoc, I understand the pain of vendoring as it's causing quite some extra work at times. However, containers/image is a library and nesting vendor folders is bad practice. In your particular case when a new dependency is added, it might work but updating existing dependencies is likely to cause errors as vendor/github.com/containers/image/vendor/dep-a might have a different API then vendor/dep-a. In other words, I don't think there is another way then the current one given the huge amount of external dependencies.

The common approach is to open another PR over a github.com/containers/skopeo that vendors in the PR's branch via the vendor.conf. If I had a branch named `foo` at `github.com/vrothberg/image` I would vendor it for Skopeo's by adding:
```
github.com/containers/image foo https://github.com/vrothberg/image
```

**Comment by Silvanoc** [comment]
@vrothberg  I understand that projects integrating `containers/image` (like `skopeo`) provide their own vendoring, as [documented in the README.md](https://github.com/containers/image#dependencies).

Perhaps my formulation of the issue isn't clear enough. I'm talking about the dependencies for testing `containers/image`. I mean, I'd expect to be able to get the tests through without touching any other repository. Of course, the integration of the patched `containers/image` into another project (e.g. `skopeo`) will require also patching that project. But that's the second step after getting `containers/image` patched.

The integration of `containers/image` into `skopeo` for the test in the [`test-skopeo`](https://github.com/containers/image/blob/0c6cc8e1420001ae39fa89d308b3b3bc5ee81c57/Makefile#L68) Makefile target should take care of "updating" the `skopeo` vendoring for the test taking into account. For example, my PR #546 is adding `github.com/coreos/pkg` and I'd like to get this new dependency "injected" into the `skopeo` dependency list only for the test.

BTW, the proposed approach wouldn't work since the test is using `master` branch of `skopeo`. It'd work testing against a patched version of `skopeo` adding `github.com/coreos/pkg`, but then my PR would be incorrect!

**Comment by vrothberg** [comment]
@Silvanoc,  I understand your desire to change the CI as I share it with you. However, I couldn't come up with a better approach than the current one (I don't think we can move c/image over to c/skopeo). I also think that the proposed "vendor.conf injection" isn't much better either as it would only work in trivial cases when the injection of new dependencies or updating existing ones does not require changes to the code.

> BTW, the proposed approach wouldn't work since the test is using master branch of skopeo.

That's correct as a PR must be opened against skopeo. It's a manual approach. The PR for c/image must pass unit tests and the PR for c/skopeo must pass everything (~integration). We face a similar situation with containers/storage.

What I think would be more friendly (and doable) is to split unit tests and the skopeo integration tests into separate Travis jobs, so it becomes clearer for developers and reviewers?

But in the end, that's just my 2 cents :) @mtrmac @runcom @rhatdan, WDYT?

**Comment by Silvanoc** [comment]
> I also think that the proposed "vendor.conf injection" isn't much better either as it would only work in trivial cases when the injection of new dependencies or updating existing ones does not require changes to the code.

The only changes on c/image requiring changes on Skopeo source code should be those changing the interfaces, what hopefully doesn't happen that often. In those cases the dependency list is the smallest issue that you will face.

WRT dependencies: I'd expect most cases to not change the dependencies at all. A couple of cases just adding new dependencies or updating the dependencies versions. And very rare cases where some dependencies are being removed.

**Comment by Silvanoc** [comment]
I've written a script to check if all dependencies reported by `c/image` are also specified in `skopeo`. The result is that the version differs in many libraries (what shouldn't be a big deal) and following two are missing:
* 

#!/bin/bash

ci_deps="vendor.conf"
skopeo_deps="/home/silvano/repositories/github/containers/skopeo/vendor.conf"

```
while read -r ci_dep ; do
    if ! echo "${ci_dep}" | grep "^#" > /dev/null ; then
        if test -z "${ci_dep}" ; then
            continue
        fi
        match_found=0
        while read -r skopeo_dep ; do
            if ! echo "${skopeo_dep}" | grep "^#" > /dev/null ; then
                if [ "${ci_dep}" = "${skopeo_dep}" ] ; then
                    #echo "Dependency found in both: ${ci_dep}"
                    match_found=1
                    continue
                else
                    lib=$(echo "${ci_dep}" | cut -d ' ' -f 1)
                    if echo "${skopeo_dep}" | grep "${lib}" > /dev/null ; then
                        echo "<> Dependency differs in the version: ${ci_dep} <=> ${skopeo_dep}"
                        match_found=1
                        continue
                    fi
                fi
            fi
        done < ${skopeo_deps}
        if test "${match_found}" -eq 0 ; then
            echo "++ Dependency missing in skopeo: ${ci_dep}"
        fi
    fi
done < ${ci_deps}

```

**Comment by Silvanoc** [comment]
@vrothberg let me try to use my real use-case to better understand the common approach.

My PR #546 adds the dependency `github.com/coreos/pkg` to `containers/image`. The tests fail because `skopeo` doesn't have that dependency.

How should I address this issue? Should I create a PR on `containers/skopeo` from the branch `master` of my repository (`silvanoc/skopeo`) and then add a last commit to my `containers/image` PR changing [`SKOPEO_REPO`](https://github.com/containers/image/blob/f0cbc16b444d729362c78244c715b45bf4019b71/Makefile#L4) to `silvanoc/skopeo`? It would imply manually removing that last commit of my PR on `containers/image` before merging my PR, right?

**Comment by vrothberg** [comment]
> How should I address this issue? Should I create a PR on `containers/skopeo` e.g. from the branch `master` of my repository (`silvanoc/skopeo`) and then add a last commit to PR #546 changing [`SKOPEO_REPO`](https://github.com/containers/image/blob/f0cbc16b444d729362c78244c715b45bf4019b71/Makefile#L4) to `silvanoc/skopeo`?

I explained the workflow above (see https://github.com/containers/image/issues/552#issuecomment-449057808). What you need to do in order to test #546 is to open a PR over at Skopeo that vendors the branch of containers/image that you used to open #546 (plus updating the dependencies), which in this specific case means to change Skopeo's `vendor.conf` to include:
```
github.com/containers/image master github.com/silvanoc/image
github.com/coreos/pkg v4
```

Once #546 is merged into container/image, the PR at Skopeo can be updated to vendor code directly from containers/image again (using the latest/specific commit).

> It would imply manually removing that last commit of PR #546 on `containers/image` before merging it, right?

You can keep it for now to wait for more feedback from other reviewers; at the moment it's just me :) It may help to collect some ideas to improve the current CI situation. I completely understand your desire to change it. Note that due to the upcoming holidays, it is likely to take until beginning of January.

**Comment by Silvanoc** [comment]
> Note that due to the upcoming holidays, it is likely to take until beginning of January.

Sure! Enjoy these days! I'm not in a hurry.

**Comment by Silvanoc** [comment]
@vrothberg I'm not 100% that I understood your explanation, but I hope so.

I've created PR #554 to fix my issue. It apparently also [doesn't create any regressions](https://travis-ci.org/containers/image/builds/472193226?utm_source=github_status&utm_medium=notification).

**Comment by mtrmac** [comment]
> However, containers/image is a library and nesting vendor folders is bad practice. In your particular case when a new dependency is added, it might work but updating existing dependencies is likely to cause errors as vendor/github.com/containers/image/vendor/dep-a might have a different API then vendor/dep-a. In other words, I don't think there is another way then the current one given the huge amount of external dependencies.

It’s not even just that — the published API of a library can’t use types from package vendored inside the library, because the outside users are completely prohibited from naming those types (in particular, consider `github.com/opencontainers/go-digest.Digest`: if it were vendored in c/image, all references to it would be mapped to `github.com/containers/image/vendor/github.com/opencontainers/go-digest.Digest` — but from the outside, `github.com/opencontainers/go-digest.Digest` and `github.com/containers/image/vendor/github.com/opencontainers/go-digest.Digest` are different and incompatible types, and Go completely prohibits using the second name at all.

So, `go-digest` _must not_ be vendored by c/image.

Still, quite a few of the dependencies of c/image do not leak through the API, so may be we could vendor some of them and make things possibly easier on callers.  But then again, almost all vending tools (at least by default?) strip the `vendor` subdirectories of vendored packages, so I’m not sure that it would help at all.

Anyway, with the proposed Go module design, this will all end up looking very different, so it seems  practical to only revisit this together with moving to the Go module use.

**Comment by mtrmac** [comment]
> Perhaps my formulation of the issue isn't clear enough. I'm talking about the dependencies for testing `containers/image`. I mean, I'd expect to be able to get the tests through without touching any other repository.

Right, this is a _perfectly reasonable_ expectation. Unfortunately, though, given how the projects have developed, the tests were written using `skopeo`, and they are the only integration / against-real-registry tests that we have, so we can’t just _drop_ them.

The Right Thing to do would be to convert almost all of the `skopeo` tests into integration tests of c/image, and then have `skopeo` perhaps only have unit tests for its CLI, and a few smoke tests, as a ~trivial wrapper over c/image. It’s just a fair amount of work.

---

> BTW, the proposed approach wouldn't work since the test is using `master` branch of `skopeo`. It'd work testing against a patched version of `skopeo` adding `github.com/coreos/pkg`, but then my PR would be incorrect!

Historically, when splitting c/image and `skopeo`, `skopeo` was the only real user and we were changing the API quite a bit, and far too often API or functionality changes were accepted into c/image that broke `skopeo` without anyone noticing — or perhaps the breakage was known but it was unanticipated that the fix would be non-trivial, and we ended up with `skopeo` broken against c/image:master for weeks.

So, the current workflow (documented in https://github.com/containers/skopeo#dependencies-management , BTW), was designed to ensure that _before_ a c/image PR is accepted, a corresponding skopeo PR is ready, reviewed, and tested to work as expected. It’s a bit of manual work and waiting for tests, and at that time it was judged very much worth it.

It’s been some time, though, API changes and broken functionality are becoming a bit rarer, so it _might_ well be time to rethink this workflow or to relax it a bit.

You are quite right that it would be nice if just adding a dependency worked smoother than it does; still, we do need to run the `skopeo` tests because they are the only ones we have.

As for how to do that, `test-skopeo` can’t just remove `skopeo/vendor` and replace it with `c/image/vendor`, because `skopeo` depends on some packages (like `urfave/cli`) that `c/image` does not use. So, we would have to somehow merge the two `vendor` directories; that does not seem _impossible_ but it’s not quite trivial either.

**Comment by mtrmac** [comment]
> What I think would be more friendly (and doable) is to split unit tests and the skopeo integration tests into separate Travis jobs, so it becomes clearer for developers and reviewers?

I don’t quite think that this is the hard part, seeing the `test-skopeo` target failure in the build log quickly points at the right problem; I guess that would be the manual work for filing the “do not merge” skopeo PR, and updating it after the c/image PR is merged, and waiting for the CI in both cases, and finding the process documentation for how to do all of this.

Maybe the `test-skopeo` target should have a comment pointing at the process documentation, at least.

But I also can’t see that splitting the two would hurt in any way (unless Travis has some limits / throttling that would make it slower overall), so if you want to do that, why not.

**Comment by Silvanoc** [comment]
One of the issues that seem to bother you is the lack of "dependency propagation", right? I mean, a new dependency in c/image not getting propagated to Skopeo and thus breaking the later integration. But what about other projects integrating c/image? For example, CRI-O. Wouldn't CRI-O be facing exactly that issue on time?

I understand that in the past being Skopeo the only tool integrating c/image the coupling seemed meaningful, but if I understand it right with CRI-O, buildah, libpod, ... that time is over.

Using Skopeo for testing c/image is perfectly acceptable from my point of view. But I cannot agree on the coupling just to ensure the dependency propagation.

I'd rather enforce c/image PRs adding dependencies to provide their counterpart PRs on Skopeo (possibly also on CRI-O, buildah, libpod, ...).

**Comment by Silvanoc** [comment]
Although I find the current approach confusing and cumbersome, I think I got it right this time and will be able to also give it a try.

**Comment by Silvanoc** [comment]
> Although I find the current approach confusing and cumbersome, I think I got it right this time and will be able to also give it a try.

Waiting for company approval for contributing to Skopeo before I can go this way :unamused:
